### ISSUE TITLE
Cisco ASAv, Netmiko (3.4.0) fails on file transfer

### ISSUE THREAD

**Issue by NMJCaspers** [ktbyers/netmiko#2650]
Using below code I have a problem to retrieve a file from the Cisco ASA (ASAv 9.14(2)15) using scp (file_transfer).

A normal SSH connection works fine, no problems at all.

But when executing the file_transfer with device_type: 'cisco_asa' some weird stuff is happening.
Looking at the logs I see an SSH session that grabs a dir.
Then I see a bunch of https connections to port 443 with various results but mostly connections being terminated?
Netmiko "thinks" the file transfer went through but it didn't.
Also, my ASDM connection is killed in the process.
Testing with ASDM disconnected results in the same problem.

For a second test I changed the device type to 'cisco_ios'.
Checking the logs, I see the first ssh session that works fine to retrieve a directory.
Next, I see a second ssh session being set up.
Authentication goes through normal; the log shows that the session is permitted.
However, the next log line is a disconnect from the server with an internal error.
This is not detected by Netmiko. It still thinks everything went okay, giving the same output.
In this case my ASDM connection stays alive.

Python Code:

```python
device = {
            'device_type': 'cisco_asa',
            'ip':       '<ip_address>',
            'username': '<username>',
            'password': '<password>',
            'port':     <port_number>,
            'secret':   <password>,
            "verbose": "True",
            'fast_cli': "False",
            'conn_timeout': 30,
            'auth_timeout': 30}
with ConnectHandler(**device) as conn:
    ft_result = file_transfer(conn, 
                direction = "get",
                verify_file = False,
                source_file = r"<filename>", 
                dest_file = r"<path>", 
                file_system = r"disk0:",
                overwrite_file = True) 
    print("result:",ft_result)
```

Output:
SSH connection established to 10.32.1.100:22
Interactive SSH session established
result: {'file_exists': True, 'file_transferred': True, 'file_verified': False}

However, no file is transferred.

**Comment by ktbyers** [comment]
This is not Netmiko i.e. must just be ASDM or something else.

```
Then I see a bunch of https connections to port 443 with various results but mostly connections being terminated?
```

What are the values of each of these?

```
                source_file = r"<filename>", 
                dest_file = r"<path>", 
                file_system = r"disk0:",
```

And then can you show me a `dir` on disk0: from the ASAv?

**Comment by NMJCaspers** [comment]
Thanks for your response. 

The source_file is backup.tar.gz and the dest_file  is e:\temp\backup.tar.gz
Just so you know I had this code working with my test ASAv around the15th of November 2021, with cisco_ios as the device type.
I don't know if 3.4.0 is newer than the version that was around at that time, but my guess is that that is the case.

ciscoasa# dir

Directory of disk0:/

9      drwx  4096         16:56:00 Nov 02 2021  smart-log
7      drwx  4096         16:54:26 Nov 02 2021  log
10     drwx  4096         16:56:04 Nov 02 2021  coredumpinfo
100    -rwx  111770516    12:46:12 Nov 03 2021  asdm-openjre-7161-150.bin
101    -rwx  72895584     13:25:36 Nov 03 2021  asa-restapi-7161-lfbff-k8.SPA
111    -rwx  50937        18:42:06 Feb 18 2022  backup.tar.gz

3 file(s) total size: 184717037 bytes
8571076608 bytes total (8215359488 bytes free/95% free)

**Comment by NMJCaspers** [comment]
I did check the file dates and it does seem that 3.4.0 is the same version as in November 2021.
I'm certain that the ASAv didn't change so I do not know why the code is misbehaving now.

**Comment by NMJCaspers** [comment]
If it may help, here's the logs from the ASAv and the SCP sessnion logs using both device types.

for obvious reasons i've obfuscated the ip addresses. host.local is my dev system, asav.local is the ASAv VM. 

[device_ios.log](https://github.com/ktbyers/netmiko/files/8101902/device_ios.log)
[device_asa.log](https://github.com/ktbyers/netmiko/files/8101903/device_asa.log)
[scp-session_ios.log](https://github.com/ktbyers/netmiko/files/8101933/scp-session_ios.log)
[scp-session_asa.log](https://github.com/ktbyers/netmiko/files/8101934/scp-session_asa.log)

**Comment by NMJCaspers** [comment]
Okay, after further debugging I have it working again for device type cisco_ios on Netmiko 3.3.3!

Using debugging I found that at first the ASA didn't have SCP enabled somehow. 
So, I enabled it manually and did a write memory to save it permanently.

This changes the result, but still doesn't solve the problem.

Now Netmiko fails (most of the time) with an exception, but the exception message itself is empty.

Then I did another debug session on the ASAv (debug ssh) and it now reported that SCP was not allowed for the account used.

I was using admin, and I did add the line conn.enable() just before the file_transfer because Netmiko was complaining about it not being there for device type cisco_ios.

However, in the sys-logs I do see a second session being created after the directory has been queried, so perhaps Netmiko's not using enable mode on the 2nd session causing the problem at this stage.

I also know admin only has privilege level 2 when initially logging in, so I tried by creating a new account that has privilege level 15. This did the trick, and the file transfer went through okay.

This doesn't take away that Netmiko is failing to see that SCP is disabled, failing to see the second session has been terminated and also fails to see that the privilege for the user was at fault.

Using the sh ssh command shows if SCP is enabled. If the output doesn't mention that SCP is enabled, then it's not.
I am now "manually" checking for that as a workaround to prevent any problem with that.

I will need to think on a way to detect that SCP is not allowed at the privilege level of the user.
Basically, the (SCP) copy command requires by default privilege level 15, but that can be changed.

For example: privilege exec level 7 copy, and possibly privilege exec level 7 copy, that'll bring it down to privilege level 7.

Possibly this can be seen in a show run, but I don't want to parse through all of that.

I will see tomorrow if I can get it to work for Netmiko 3.4.0, although I'm fairly confident that it will work.
Then I will also see if it is working the same for device type cisco_asa, first on 3.3.3 and then on 3.4.0.

**Comment by ktbyers** [comment]
So you are using the `cisco_ios` device_type and it works, but it fails with the device_type as `cisco_asa`?

You have to have a privilege 15 account directly, you can't call `enable()` on the secure copy channel (i.e. you are elevating privileges on the SSH control channel, but on the SCP channel you can't elevate privileges). It has to be a direct privilege 15 account or more exactly the necessary privileges for that SCP operation have to be granted to the user directly (i.e. without any enable call).

`This doesn't take away that Netmiko is failing to see that SCP is disabled`

Typically this error is pretty self-evident, but it is possible it is more obscure on the ASA. So feel free to submit a PR to improve the error handling here. Netmiko will not check the SSH status by parsing configs or show commands (this is intentional). It will just try to do the SCP and report the failure.

`failing to see the second session has been terminated and also fails to see that the privilege for the user was at fault.`

That is a bit harder, but you are welcome to work on improving the error handling there as well. 

Netmiko wouldn't do any parsing of configs, or user privilege levels (i.e. I would reject a PR on that). Once again Netmiko will just try to do the SCP and report the failure, but it is definitely possible this lower level message is confusing and could be improved.

**Comment by NMJCaspers** [comment]
Thanks for your response. It's clear and understandable that you don't want to be parsing configuration in the library.
Perhaps a demo and/or documentation could cover that part.

As for a pull request, the code in the library is so complex that it exceeds my capability of following the flow and see what is happening. I always treat libraries as black boxes.  I adhere to the API that is in the documentation and/or examples and then I expect some outcome. It's either an error message that is clearly stating what the problem is or it's the result that I expected.

Calling conn.enable() is a must_have for device type cisco_ios, even though I'm using a privilege 15 account.
Without it, it will consistently fail the download and throw an exception:
Invalid output from MD5 command:           ^
ERROR: % Invalid input detected at '^' marker.

However, with this in place the result of the download looks to be rock solid.
I've done tens of consecutive tests and they all succeeded as expected.

If I however switch to cisco_asa as a device type, the results become inconsistent.
The download itself succeeds, but most of the time an exception occurs.
Most of the time the exception message itself is blank.
Very rarely sometimes also this exception occurs somewhere inside the library:
Socket exception: An established connection was aborted by the software in your host machine (10053)
This exception is printed and is handled.
It is not the exception that gets caught by my outer exception block, the outer exception block does get an exception, but again, it has an empty message.
(I'm printing the extra '@ ' to be able to differentiate what messages come from my code and what messages originate from elsewhere.)

Having switched to Netmiko v3.4.0 did not change anything about that behavior.
For now, I will keep using cisco_ios as device type as that seems to be the more reliable approach.

Below I've pasted my test code to reproduce issues:

def ScpGet(self, source: str, destination: str, verbose: bool = False):
        device = {
            #'device_type': 'cisco_ios',
            'device_type': 'cisco_asa',
            'ip':       self.HostIP,
            'username': self.Username,
            'password': self.Password,
            'port':     22,
            'secret':   self.Enable,
            'conn_timeout': 30,
            'auth_timeout': 30,
            'session_log': 'scp-session.log',
        }
        
        try:
            with ConnectHandler(**device) as conn:
                if verbose:
                    print("@ Scp downloading file.")
                if device['device_type'] == 'cisco_ios':
                    conn.enable();
                ft_result = file_transfer(conn, 
                    direction = "get",
                    verify_file = False,
                    source_file = source, 
                    dest_file = destination, 
                    file_system = r"disk0:",
                    overwrite_file = True) 
                if verbose:
                    print("@",ft_result);
                return True;
        except Exception as ex:
            if verbose:
                print("@ Exception connecting/downloading backup:", ex);
            self.lastError = ex;
            return False;

**Comment by ktbyers** [comment]
Can you get rid of your try/except statement and send over the full exception stack trace for a failure using the `cisco_asa` device_type.

**Comment by NMJCaspers** [comment]
It looks like this is a Paramiko error, not a Netmiko problem.

Exception stack trace:

>python netmiko_asa_test.py
Traceback (most recent call last):
  File "netmiko_asa_test.py", line 16, in <module>
    ft_result = file_transfer(conn,
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\netmiko\scp_functions.py", line 134, in file_transfer
    verifyspace_and_transferfile(scp_transfer)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\netmiko\scp_functions.py", line 39, in verifyspace_and_transferfile
    scp_transfer.transfer_file()
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\netmiko\scp_handler.py", line 365, in transfer_file
    self.get_file()
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\netmiko\scp_handler.py", line 370, in get_file
    self.scp_conn.scp_get_file(source_file, self.dest_file)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\netmiko\scp_handler.py", line 57, in scp_get_file
    self.scp_client.get(source_file, dest_file)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\scp.py", line 275, in get
    self.close()
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\scp.py", line 287, in close
    self.channel.close()
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\paramiko\channel.py", line 671, in close    self.transport._send_user_message(m)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\paramiko\transport.py", line 1917, in _send_user_message
    self._send_message(data)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\paramiko\transport.py", line 1893, in _send_message
    self.packetizer.send_message(data)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\paramiko\packet.py", line 431, in send_message
    self.write_all(out)
  File "%appdata%\Local\Programs\Python\Python39\lib\site-packages\paramiko\packet.py", line 367, in write_all
    raise EOFError()
EOFError


Testcode:

from netmiko import ConnectHandler, file_transfer

device = {
    'device_type': 'cisco_asa',
    'ip': "<device ip/host>",
    'port':22,
    'conn_timeout': 30,
    'auth_timeout': 30,
    'username': "<user>",
    'password': "<user password>",
    'secret': "<enable password>",
    'session_log': 'scp-session.log'
}

with ConnectHandler(**device) as conn:
    ft_result = file_transfer(conn, 
        direction = "get",
        #verify_file = False,
        source_file = "backup.tar.gz", 
        dest_file = "c:\\temp\\new\\backup.tar.gz", 
        file_system = r"disk0:",
        overwrite_file = True)
    

SCP Session Log:
User <username> logged in to ciscoasa
Logins over the last 1 days: 176.  Last login: 23:11:39 CEST Feb 20 2022 from 10.32.1.1
Failed logins since the last login: 0.  
Type help or '?' for a list of available commands.
ciscoasa> show curpriv
Username : <username>
Current privilege level : 1
Current Mode/s : P_UNPR
ciscoasa> 
ciscoasa> 
ciscoasa> 
ciscoasa> 
ciscoasa> 
ciscoasa> enable
Password: ****
ciscoasa# 
ciscoasa# terminal pager 0
ciscoasa# 
ciscoasa# configure terminal
ciscoasa(config)# 
ciscoasa(config)# terminal width 511
ciscoasa(config)# 
ciscoasa(config)# end
ciscoasa# 
ciscoasa# 
ciscoasa# 
ciscoasa# 
ciscoasa# 
ciscoasa# 
ciscoasa# verify /md5 disk0:/backup.tar.gz
!!!!!!!!!!!!!Done!
verify /MD5 (disk0:/backup.tar.gz) = 920c3b5ae474f9cb6df86deac39d176c
ciscoasa# 
ciscoasa# 
ciscoasa# 
ciscoasa# dir disk0:/backup.tar.gz

Directory of disk0:/backup.tar.gz

96     -rwx  50987        19:00:50 Feb 20 2022  backup.tar.gz

1 file(s) total size: 50987 bytes
8571076608 bytes total (8214315008 bytes free/95% free)

ciscoasa# 
ciscoasa# exit

**Comment by ktbyers** [comment]
I just read that as...you tried to do a `get` and then something closed Secure Copy session.

So cisco_ios works using the same privilege 1 user?

**Comment by NMJCaspers** [comment]
This seems a "feature" on the Cisco end, as the user account used is a privilege level 15 account.
Any account with a lower privilege level will fail the file transfer.

**Comment by ktbyers** [comment]
Yeah, you need a user that has direct privileges to perform the secure copy and to write to the file system. Remember on the Secure Copy Channel itself there is no way to elevate privileges (that I am aware of).
